Type checking + refactoring for the coverage data model

latk opened the pull request:
As a “dividend” from dropping support for Python 3.6, various improvements have become possible. This PR focuses on adding more type annotations for the `gcovr.coverage` data model, externalizing the merging logic, and replacing the increasingly large tuples from `get_global_stats()` and so on with more useful classes. In doing so, a number of bugs were found, but not necessarily fixed as part of this PR.

For reviewing this PR, moving one commit at a time is recommended. The commit messages provide further explanation. Here, I summarize the central changes and consequences.

## `gcovr.coverage` must be a leaf module

Python doesn't handle cyclic dependencies very well. Since the coverage data model now provides types that are used in type annotations throughout the code, the `gcovr.coverage` module now cannot have dependencies on other gcovr modules.

In particular, the `CovData` type alias describes the top-level filename → FileCoverage dictionary.

As part of adding type annotations, some dead code regarding function coverage could be removed.

## New `CoverageStat`, `DecisionCoverageStat`, `SummarizedStats` types

The functions like `file.branch_coverage()` now return a `CoverageStat(covered, total)` object. This has a `percent` property that replaces the need to call `calculate_coverage()`. In case a default value other than None is desired, the `percent_or(default)` method can be used.

Since decision coverage has an additional `uncheckable` count, a separate but otherwise equivalent `DecisionCoverageStat` type is needed.

The `SummarizedStats` type aggregates line, branch, function, and decision coverage for a file or for the entire project, depending on which constructor function is used. This replaces the `get_global_stats()` and `summarize_file_coverage()` functions.

As a benefit of this, the different writers can now share common logic for aggregating coverage statistics, instead of doing everything themselves. Updating the writers uncovered some potential bugs, which are described below but not yet fixed.

## Merging coverage data with `merged = merge_file(left, right)` instead of `left.update(right)`

Previously, coverage data structures were merged with the `update()` method. This had a number of issues:

* It can't handle the different `DecisionCoverage` types reasonably.
* Mutable state obscures the data flow in some parts of the code.
* It may become necessary to make the merging logic dependent on command line options (perhaps for #586), but this won't work well if the merging logic is part of the `gcovr.coverage` module (as it must now be a leaf module).

To support this more “functional” style, some functions in other parts (`__main__` and `gcov`) were updated to return CovData instead of updating it in-place.

To insert individual elements, the merging module also provides functions like `insert_line_coverage(file, line)`.

Auditing the merging logic found a bug in the interplay of `--add-tracefile` + branch coverage and in the decision coverage analysis.

## Fixed bug: correctly merge branch coverage data from `--add-tracefile`

In BranchCoverage, the `fallthrough` and `throw` flags used to be tri-state booleans None/False/True where `None` meant “unknown” and `True` meant “known and set”. But the JSON writer coerced those to boolean False/True. This caused a problem because the old merging logic only checked whether the flag was None, and if so choose the other value. This meant the order of updates would become relevant when merging data loaded via `--add-tracefile`. In pseudocode:

```python
>>> BranchCoverage(throw=False).update(BranchCoverage(throw=True)
BranchCoverage(throw=False)  # keep left value

>>> BranchCoverage(throw=True).update(BranchCoverage(throw=False)
BranchCoverage(throw=True)  # keep left value
```

The correct result in either case would be `throw=True` because it should override the “unknown” state.

Now the flags use False for “unknown” and True for “known and set”, which allows the JSON output to remain stable.

## Open bug: package-level function coverage in Cobertura is entirely wrong

In pseudocode, the problem is that the Cobertura writer does its own coverage aggregation, but just overwrites instead of adding up function coverage stats:

```python
for package in packages:
  function_stat = CoverageState.new_empty()
  for file_cov in package.files:
    function_stat = file_cov.function_coverage()  # should be "+="
  report(function_stat)
```

## Open bug: branch coverage also counts noncode lines

For calculating line coverage, only the following lines are counted:

```python
if line.is_covered or line.is_uncovered: ...
```

Notably, this excludes `noncode` lines.

The `branch_coverage()` method has no such check, and will also consider branches on noncode lines!

This shouldn't matter since the gcovr_parser tries to check whether a line should contain branches, but there seems to be some discrepancy. Adding the noncode check in the coverage data model would cause a lot of test failures. Further investigation is needed to understand what's going on.

## Open bug: decision analysis can resurrect excluded lines

The decision analysis accesses coverage data via the `file.line(lineno)` method (before this PR) / `get_or_create_line_coverage(file, lineno)` (after this PR). But if no LineCoverage data structure already exists for that line, it will be created. The default LineCoverage also sets noncode=False, so that the newly-created lines will be counted as uncovered.

I haven't created a testcase to check this, but fixing this issue and moving to read-only access seems to break the tests, so correct behavior should be determined in a separate issue.

## Surprising discrepancy: null coverage in JSON-Summary writer

A never-ending problem is the default value if a coverage percentage is NaN. In the JSON-Summary writer, the project-summary and file-summary percentage fields seem to have different defaults. Whereas the project-summary defaults to 0%, the file-summary renders as null. I've kept this behaviour, for now, but it would probably make sense to figure out an unified approach eventually.

This kind of discrepancy was much easier to find when using the unified `CoverageStat.percent` method and when using a type checker such as MyPy.

## Obstacles to more type checking

At this point, more linting and more type-checking is not directly possible because there are still tons of spurious errors. In particular, heterogeneous dicts are tricky, e.g. if one dict entry contains a list and another a string. Our HTML writer is full of those. Eventually, these dicts can be replaced with dataclasses, but that is out of scope for this PR. Another problem is the Options type in the main function, since some fields change their types during execution (e.g. compiling filters). I do not intend to add MyPy to the CI process anytime soon, but I would like fewer type errors in order to have a better development experience, especially for people using IDEs or LSP-enabled editors like VS Code.

Spacetown commented on gcovr/writer/txt.py:
```diff
@@ -21,26 +21,30 @@
+        title_covered = "Taken" if options.show_branch else "Exec"
+        title_missing = "Missing"
+        fh.write(
+            "File".ljust(40)
+            + title_total.rjust(8)
+            + title_covered.rjust(8)
```
Create global constants for column length and use them also in the line for the files.

Spacetown commented on gcovr/writer/json.py:
```diff
@@ -112,64 +112,41 @@ def print_json_summary_report(covdata, output_file, options):
-                "branch_total": branch_total,
-                "branch_covered": branch_covered,
-                "branch_percent": branch_percent,
-                "function_total": function_total,
-                "function_covered": function_covered,
-                "function_percent": function_percent,
+                **_summary_from_stats(SummarizedStats.from_file(covdata[key]), None),
             }
         )
```
What about adding optional `**kwargs` to `_summary_from_stats` which are added to the dict. Than he copy of the data for each file isn't needed anymore.

latk replied:
I wouldn't expect that to affect performance noticeably. There is only going to be one copy per file in the output, which is very efficient compared to other parts of the code. This is Python, after all. I would totally agree with you if this were C++ or Rust, though :) 

I think flattening the return value from this function makes the data flow clearer than routing parts of the dict through the function.

Spacetown reviewed (commented):
LGFM, only some questions.

latk commented on gcovr/coverage.py:
```diff
@@ -400,6 +411,9 @@ def decision_coverage(self) -> Tuple[int, int, int, Optional[float]]:
         return total, cover, unchecked, percent
 
 
+CovData = Dict[str, FileCoverage]
```
The `covdata` has always been a dict, and I didn't intend to change that as part of this PR. But it would definitely be possible to introduce a `ProjectCoverage` class or something that.

What do you mean with debug output?

Spacetown replied:
Output like `Adding coverage for file...`

Also there where two places which add a new file coverage object if file isn‘t in dict. This can then be done while accessing the file the first time.

latk replied:
I see your point.

Ok, I'll think about this and experiment a bit. But if I can't find a good design I'd rather keep the status quo and maybe revisit that point in the future. I'm trying to balance “whee fun refactoring” with “one step at a time, shipping something *good* is better than waiting until everything is *perfect*”.

One concern I have with adding such logging is that there is no clear “first” time. For example, consider the case of multiple worker threads. Initially, these load coverage data into separate CovData dicts. Thus, multiple workers might see the same file for the “first” time, if the same file is part of multiple compilation units. Doing this correctly would require something like a shared set of known filenames protected by a mutex, or a channel in which workers can publish file-processing events that are consumed by a separate thread. But that would be behaviour that is out of scope of the coverage data model.

Spacetown replied:
We can also finish this PR and do this later. After merging this and #597 we can fix the two bugs before creating the release.

latk commented on gcovr/merging.py:
```diff
@@ -0,0 +1,202 @@
+        return right
+
+    # Merge Conditional decisions.
+    Conditional = DecisionCoverageConditional
+    if isinstance(left, Conditional) and isinstance(right, Conditional):
```
Yes, only to avoid awkward linebreaks. I can pull the variables to the top and add a comment.

Spacetown replied:
What about using DC or DesCov instead of DecisionCoverage in all places?

latk commented:
I've rebased and worked through the review remarks.

I thought about replacing the CovData type alias with a proper class, but found myself moving in the other direction – moving more behaviour away from the coverage data model and into the new `gcovr.merging` module or other places (for example, some formatting routines are now moved into the txt writer).

The `file.line(lineno)` method looks like an accessor, but is actually a mutator: it will create and insert a default line if it doesn't already exist. This behaviour was not clear and leads to problems. I've replaced those uses with an `insert_line_coverage(file, line)` function where that behaviour was unnecessary, and a `get_or_create_line_coverage(file, lineno)` function where I wanted to retain bug-compatibility (for now).

This discovered some additional problems, in particular with the decision analysis: I think it can resurrect coverage information for lines that were previously excluded.

Spacetown reviewed (approved):
LGFM

Spacetown commented on gcovr/gcov_parser.py:
```diff
@@ -613,9 +626,11 @@ def _add_coverage_for_function(
+    insert_function_coverage(
+        coverage,
+        FunctionCoverage(name, lineno=lineno, call_count=calls),
+        MergeOptions(ignore_function_lineno=True),
```
Do you really want to ignore the line number? As I mentioned in https://github.com/gcovr/gcovr/issues/586#issuecomment-1086693394 I prefer to wait if it can be solved on project level.

latk replied:
Yes.

My proposed implementation is backwards- (or rather: bug-) -compatible. It maintains exactly the existing behaviour. That is:

* While parsing a .gcov file, the function is attributed to the last line on which it occurs. Therefore, the ignore-function-lineno option is set in this code snippet. This is the equivalent to the old code, which *overwrites* the lineno field:

  ```python
  function_cov = coverage.function(name)
  function_cov.lineno = lineno
  function_cov.count += calls
  ```

* But when merging .json files, this option is not set and the problematic assert in #586 may be triggered.

(To be very pedantic, there is a slight change in behaviour for the .gcov parsing case: The new code uses `max(left.lineno, right.lineno)` to maintain symmetry, whereas the old code just overwrites with the current `lineno`. The current lineno is not guaranteed to increase monotonically due to the presence of template specialization sections. But I'm not sure if that can actually lead to a different outcome.)

Once this PR is merged, I'll think about whether and how #586 can be addressed. Maybe the solution is to also enable this option when merging FileCoverage (on an opt-in basis with a command line flag), maybe the solution is to add a MergeOption to change the concept of function-identity from `name` to `(name, lineno)`. And maybe this is something that can't be supported at all in gcovr.
